crates/bevy_ecs/src/component/mod.rs

```diff
@@ -290,9 +289,52 @@ impl Components {
     }
 }
 
-bitflags! {
-    pub struct ComponentFlags: u8 {
-        const ADDED = 1;
-        const MUTATED = 2;
+#[derive(Copy, Clone, Debug)]
+pub struct ComponentCounters {
+    added: u32,
+    changed: u32,
+}
+
+impl ComponentCounters {
+    #[inline]
+    pub fn is_added(&self, system_counter: u32, global_system_counter: u32) -> bool {
+        let component_age = global_system_counter.wrapping_sub(self.added);
+        let system_age = global_system_counter.wrapping_sub(system_counter);
+
+        component_age < system_age
+    }
+
+    #[inline]
+    pub fn is_changed(&self, system_counter: u32, global_system_counter: u32) -> bool {
+        let component_age = global_system_counter.wrapping_sub(self.changed);
+        let system_age = global_system_counter.wrapping_sub(system_counter);
+
+        component_age < system_age
+    }
+
+    pub(crate) fn new(global_system_counter: u32) -> Self {
+        Self {
+            added: global_system_counter,
+            changed: global_system_counter,
+        }
+    }
+
+    pub(crate) fn check_counters(&mut self, global_system_counter: u32) {
+        check_counter_impl(&mut self.added, global_system_counter);
+        check_counter_impl(&mut self.changed, global_system_counter);
+    }
+
+    #[inline]
+    pub(crate) fn set_changed(&mut self, global_system_counter: u32) {
+        self.changed = global_system_counter;
+    }
+}
+
+fn check_counter_impl(counter: &mut u32, global_system_counter: u32) {
+    let counter_age = global_system_counter.wrapping_sub(*counter);
```

Comment:
For the added counter, it's the number of counts since the creation of the component, which is why I called it "age". For changed counters, it's only since the last change, which is probably more of an "age difference". I like the clarity of saying: "a change/addition is detected when the system is older than the component".
I don't have a strong preference there, what do you want to call it?

Reply:
I prefer age difference over age in the latter case.

Reply:
Renamed to "delta" in ae3f3aa1e3fbfb05b9f3506a8abc5ce257db8c71 to be consistent with previous renames.